linkerd/main/src/main/scala/io/buoyant/Linkerd.scala

```diff
@@ -73,7 +52,77 @@ object Linkerd extends App {
         if (!f.isFile) throw new IllegalArgumentException(s"config is not a file: $path")
         Source.fromFile(f).mkString
     }
-
     Linker.parse(configText)
   }
+
+  private def initAdmin(linker: Linker, linkerConfig: LinkerConfig): Closable with Awaitable[Unit] = {
+    val linkerdAdmin = new LinkerdAdmin(this, linker, linkerConfig)
+    val adminInitializer = new AdminInitializer(linker.admin, linkerdAdmin.adminMuxer)
+    adminInitializer.startServer()
+    adminInitializer.adminHttpServer
+  }
+
+  private def initRouter(config: Router): Closable with Awaitable[Unit] = {
+    val router = config.initialize()
+
+    val servers = router.servers.map { server =>
+      log.info("serving %s on %s:%d", server.router, server.ip, server.port)
+      val listening = server.serve()
+      val announcements = announce(router.announcers, server)
+      Closable.sequence(announcements, listening)
+    }
+
+    new Closable with CloseAwaitably {
+      private[this] val closer = Closable.sequence(Closable.all(servers: _*), router)
+      def close(deadline: Time) = closeAwaitably { closer.close(deadline) }
+    }
+  }
+
+  private def announce(announcers: Seq[(Path, Announcer)], server: Server.Initializer): Closable =
+    Closable.all(server.announce.map(announce(announcers, server, _)): _*)
+
+  private def announce(
+    announcers: Seq[(Path, Announcer)],
+    server: Server.Initializer,
+    name: Path
+  ): Closable =
+    announcers.filter { case (pfx, _) => name.startsWith(pfx) } match {
+      case Seq() =>
+        log.warning("no announcer found for %s", name.show)
+        Closable.nop
+
+      case announcers =>
+        val closers = announcers.map {
+          case (prefix, announcer) => announce(prefix, announcer, server, name)
+        }
+        Closable.all(closers: _*)
+    }
+
+  private def announce(
+    prefix: Path,
+    announcer: Announcer,
+    server: Server.Initializer,
+    name: Path
+  ): Closable = {
+    log.info("announcing %s as %s to %s", server.addr, name.show, announcer.scheme)
+    val f = announcer.announce(server.addr, name.drop(prefix.size))
+
+    // If we close before the announcer registers, we
+    // cancel the registration. If we close after
+    // registration, we close it.
+    //
+    // XXX we should change the announcer API to return
+    // an Announcement synchronously (that may actually
+    // wrap some asynchronous announcement logic).
+    val closeRef = new AtomicReference[Closable](Closable.make { _ =>
+      f.raise(Failure("closed").flagged(Failure.Interrupted))
```

Comment:
I think we want to signal cancellation in case the announcer is out to lunch / etc.  I added a var that signals to pending-handling whether cancellation has already occurred.   It's not 1000% safe, but I think it's basically as good as we can dod.